### PR TITLE
Matthewcurtis1999/vir 1904 attributeerror nonetype object has no attribute pop

### PR DESCRIPTION
**Pull request by mattcurts**
Added return type hints to all functions that have @emits, and I think the issue is that ``get_json()`` which gets assigned to ``received`` is a shortcut for ``get()`` is either the message received or None signifying that the channel has been unsubscribed and no more messages will be received.

Added a logging message and then subscribe to the events channel before continuing in the loop, waiting for the next event.

https://aioredis.readthedocs.io/en/v1.3.0/api_reference.html

### CODE REVIEW

**Comment by mattcurts** [comment]
no clue why CI is flipping the order of test snapshot

File: virtool/data/events.py [hunk 1]
```diff
@@ -203,6 +203,13 @@ async def __anext__(self) -> Event:
             try:
                 received = await self._channel.get_json()
 
+                if received is None:
+                    logger.error(
+                        "Channel has been unsubscribed, trying to re-subscribe"
+                    )
+                    (self._channel,) = await self._redis.subscribe(self._channel_name)
+                    continue
+
```
Comment (igboyes) [inline virtool/data/events.py:212]: Remove this. We are already handling `ChannelClosedError` more rigourously below.

This problem here is that `None` is being put in the queue. Update the `emit` function such that it avoids ever putting `None` in the queue. Log a warning if it is attempted so we can fix the problem.

**Review by igboyes — CHANGES_REQUESTED** [review]
The `emit` function needs to avoid ever putting `None` in the queue. Update this function so it logs a warning and does nothing if `None` is received.

File: virtool/data/events.py [hunk 2]
```diff
@@ -92,6 +92,10 @@ def emit(data: BaseModel, domain: str, name: str, operation: Operation):
 
 
     """
+
+    if data is None:
+        logger.warning("emit event with no data")
+
```
Comment (igboyes) [inline virtool/data/events.py:98]: Need an `else` here so that nothing is emitted in warning case.

**Review by igboyes — CHANGES_REQUESTED** [review]

**Review by igboyes — APPROVED** [review]

**Pull request merged**